### PR TITLE
PERF: Additional performance updates

### PR DESCRIPTION
**Pull request by greglucas**
This is whittling down the last two easy to find performance bottlenecks using snakeviz. See the image below.

image

Commit 1: Removes the __hash__ calls to enum from the dict access. It is small enough and only used in one place that we can inline it.
Commit 2: (far right purple box is `ProjError.clear()`) Removes the class attribute of the internal_proj_error and moves that into Cython to take care of all the error messages without needing to access Python objects while within C.

### CODE REVIEW

File: pyproj/_datadir.pyx [hunk 1]
```diff
@@ -19,7 +19,8 @@ _USE_GLOBAL_CONTEXT = strtobool(os.environ.get("PYPROJ_GLOBAL_CONTEXT", "OFF"))
 # static user data directory to prevent core dumping
 # see: https://github.com/pyproj4/pyproj/issues/678
 cdef const char* _USER_DATA_DIR = proj_context_get_user_writable_directory(NULL, False)
-
+# Store the message from any internal PROJ errors
+cdef str _INTERNAL_PROJ_ERROR = None
```
Comment (greglucas) [inline pyproj/_datadir.pyx:23]: I'm not sure how this behaves across threads compared to the previous implementation of using class attributes to store the information. Do I need to grab a context here when using this?

Comment (snowman2) [reply pyproj/_datadir.pyx:23]: It looks like the type is string, so I imagine that it still uses the Python gil. Due to that, I believe that the behavior here should be consistent with the previous behavior.

**Comment by snowman2** [comment]
This is great! How much does it improve the performance?

**Comment by greglucas** [comment]
I go from about 1.5 us -> 1.1 us / iteration of transform_point (rebased on your branch). Maybe see the results of calling it 2,000,000 times to get the cumulative. Here are the figures below relative to the original one included above, although I'm not sure we really get _that_ much of a speedup.

After the first commit:
image

After the second commit:
image

File: pyproj/_datadir.pyx (hunk 1 above)
Comment (greglucas) [reply pyproj/_datadir.pyx:23]: Another thing to think about is whether you want `get_proj_error()` and `clear_proj_error()` to leak into the public modules or whether you want those kept as private functions.

Comment (snowman2) [reply pyproj/_datadir.pyx:23]: I think that keeping them private is a good idea.

**Comment by snowman2** [comment]
Every little speedup counts. I think that this is a worthwhile change. Thanks for working on this 👍

File: pyproj/exceptions.py [hunk 2]
```diff
@@ -1,27 +1,25 @@
...
-            )
-            ProjError.clear()
+        proj_error = _get_proj_error()
+        if proj_error is not None:
+            error_message = f"{error_message}: (Internal Proj Error: {proj_error})"
+            _clear_proj_error()
         super().__init__(error_message)
 
     @staticmethod
     def clear() -> None:
```
Comment (snowman2) [inline pyproj/exceptions.py:18]: I think you can safely remove this method.

Comment (snowman2) [reply pyproj/exceptions.py:18]: Actually, I think it is fine to ignore that in this PR.

**Pull request merged**